Nonpreferred option

huntc opened the pull request:
This DeferredResult design is a simple re-factoring of CompletableFuture where by CompletableFuture's non-completion methods are removed.

Specific concerns with this interface as per CompletableFuture:
- it is not consistent with the Java 8 stream API. 
- method names such as thenApply and thenCompose are different to names used by the stream API such as map and flatMap. 
- CompletableFuture has a thenApply functor method, a thenCompose monad method, methods for handling failures, and methods for completing the future successfully or with an exception. 
- There are a number of useful helper methods that it provides, but all of these could be achieved using the methods already mentioned. 
- Methods such as thenApply have three versions: one that gets invoked synchronously by the completing thread or the calling thread if already completed; and then two “async” versions where the callbacks are invoked in the default or in a supplied execution context.
- Since CompletableFuture is not an interface it can only have one implementation. Other APIs that provide futures (Scala, Netty, Guava) cannot be wrapped by this interface, meaning that it will not play well with the rest of the ecosystem, and is therefore not a good candidate for being returned or accepted by an API.
- CompletableFuture leaks its internals including methods such as getNumberOfDependents and obtrudeValue. This leakage makes CompletableFuture appear as a low level concurrency construct and not an API abstraction that can be shared between libraries.

huntc committed:
Initial commit of a second approach whereby we've extracted all non-completing methods from CompletableFuture and removed the inheritance from Future. Still need to consider whether to support JDK 6 (presently relies on JDK 8).

huntc committed:
Merge branch 'master' into nonpreferred

Conflicts:
	src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java

huntc committed:
Re-adding

viktorklang commented on src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,799 @@
+     *          if this future completed
+     *          exceptionally or a completion computation threw an exception
+     */
+    T join();
```
This shouldn't be here, it's blocking

huntc replied:
Will be removed.

viktorklang commented on src/main/java/com/typesafe/scalacompat/concurrent/DeferredResult.java:
```diff
@@ -0,0 +1,799 @@
+     *          if this future completed
+     *          exceptionally or a completion computation threw an exception
+     */
+    T getNow(T valueIfAbsent);
```
This shouldn't be here, it's blocking.

huntc replied:
It isn't blocking.

viktorklang replied:
You're right. I guess a better method would be `def value: Optional[T]`

huntc replied:
Yeah, we should probably stick to the existing interface though.

viktorklang replied:
Alright, deal

huntc committed:
Mostly applied JDK 8 features

viktorklang commented:
What I'd like to see here now in this PR is a retrofitted version of CompletableFuture implementing DeferredResult, to show that it works.

huntc committed:
Now includes a CompletableFuture that implements DeferredResult

huntc committed:
Added in the JDK test written for CompletableFuture. Tests can be run using the following sbt command:

sbt "test:run-main Basic"

huntc committed:
Worked around UNSAFE issues (thanks Viktor!) and got the basic test working

huntc committed:
Removed the unconditional casting, and supported the ability to compose any type of deferred result

huntc committed:
1. factory methods should return the type they are declared in
2. static factory methods should appear on DeferredResult and delegate to CompletableFuture, returning DeferredResult
3. other static methods on CompletableFuture should have a DeferredResult counterpart that compose DeferredResult objects as CompletableFutures and then call the corresponding CompletableFuture static method.
4. Don't cast DeferredResult to CompletableFuture - compose instead.
5. Re-applied the Basic.java test so that it dropped casting and used DeferredResult predominanlty.

huntc committed:
Removed unwanted import

huntc committed:
Avoid passing java to scalar as it isn't yet entirely up to scratch on jdk8

viktorklang commented:
All methods returning DeferredResult in src/main/java/com/typesafe/java/util/concurrent/CompletableFuture.java should return CompletableFuture

huntc committed:
Removed unnecessary creational methods

huntc committed:
Removed another creational method from the interface

huntc commented:
No longer relevant
